File: files/400-499/402/CH-402.py
import pandas as pd
import numpy as np

path = "400-499/402/CH-402 Matrix Normalization.xlsx"
input = pd.read_excel(path, usecols="C:F", nrows=4, skiprows=3, header=None).to_numpy()
test = pd.read_excel(path, usecols="J:M", nrows=4, skiprows=3, header=None).to_numpy()

vectorized_input = input - np.diag(input)[:, np.newaxis]
print((vectorized_input == test).all())
# True

File: files/400-499/402/CH-402.R
library(tidyverse)
library(readxl)

path <- "400-499/402/CH-402 Matrix Normalization.xlsx"
input <- read_excel(path, range = "C4:F7", col_names = FALSE) %>% as.matrix()
test <- read_excel(path, range = "J4:M7", col_names = FALSE) %>% as.matrix()

normalized_input <- sweep(input, 1, diag(input), "-")
all.equal(normalized_input, test)
#> [1] TRUE
